### PR TITLE
feat(test/int): Run int tests only on pushes to master

### PR DESCRIPTION
**Pull request by german-muzquiz**
There are two concerns when running integration tests to block PR builds:
1. Builds take a long time to complete. With the few tests added so far in clouddriver, build times were increased by about 13 minutes. This is only going to take longer as more tests are added to the service.
2. Test flakiness. Given the inherent complexity of integration tests, there could be sometimes when they fail in a particular execution without any code change.

The goal of integration tests is to find bugs before doing releases, the sooner the better. A good compromise would be to run them on pushes to `master` branch, so that PR builds are not delayed or blocked by the tests, but yet there's a clear indication of exactly which PR broke the tests, which is not possible when they are run on a scheduled basis like the current end to end tests.

This PR introduces a new workflow for running integration tests only on pushes to `master` branch. It makes use of gradle's [build cache](https://docs.gradle.org/current/userguide/build_cache.html) and github's [cache action](https://docs.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows), so that only gradle tasks that were affected by a commit are executed. This allows to save a lot of time and unnecessary test runs on changes that could not affect test output. As an example, build time for changing a workflow file goes down from ~23 minutes to ~3 minutes, because gradle doesn't detect any changes and marks all tasks as up-to-date.

### CODE REVIEW

**Comment by german-muzquiz** [comment]
cc @allisaurus @maggieneterval @jonsie

**Comment by maggieneterval** [comment]
> A good compromise would be to run them on pushes to master branch, so that PR builds are not delayed or blocked by the tests, but yet there's a clear indication of exactly which PR broke the tests

I think this is a reasonable compromise, although I don't love that in the case where regressions are introduced, the burden of identifying them is essentially being pushed from the PR author to whoever is on build rotation (no worse than the current end-to-end tests, though, and as you mention, better because we can at least identify the exact PR). I don't have a better suggestion in mind, but might be worth calling out this decision in the #build-process channel as I'm not sure build rotation members are very actively monitoring the build statuses page in addition to the Jenkins server.

> It makes use of gradle's build cache and github's cache action, so that only gradle tasks that were affected by a commit are executed. 

Is it feasible use these on PR build as well?

**Comment by german-muzquiz** [comment]
> Is it feasible use these on PR build as well?

Yes! cache can also be used in PR builds to compile only the code that was changed instead of the whole project.

I'm thinking that running integration tests can also be done as part of a PR status checks, just a separate one that doesn't block the merge. Then the approver could see if it's ok to merge, wait, etc. I'll ask feedback in #build-process

File: .github/workflows/integration_tests.yml
Comment (allisaurus) [inline .github/workflows/integration_tests.yml:31]: 🔥

File: .github/workflows/integration_tests.yml
Comment (jonsie) [inline .github/workflows/integration_tests.yml:6]: I think we possibly could run on this on PR so long as it's not a build the blocks merge.

Comment (maggieneterval) [reply .github/workflows/integration_tests.yml:6]: In addition to running on PR and master builds, should we run on release branch builds as well?

Comment (german-muzquiz) [reply .github/workflows/integration_tests.yml:6]: I'm fine running or not running in release branch builds.
As long as every commit to release branches is done through PRs, the tests will run in those PRs checks before the merge. 
Moreover, not running in release branches will make that the full test suite is executed at least once in each PR regardless of any code change, because there will be no cache to inherit from. 

In the case of `master` its PRs will inherit the cache that was saved in `master` runs, at least only on the first commit in the PR. The logic is literally "get the cache for this commit sha, if not found, get the cache from the previous commit sha (if exists)".

File: .github/workflows/integration_tests.yml
Comment (allisaurus) [inline .github/workflows/integration_tests.yml:34]: How are we specifying that this check is non-blocking to PRs? If that has to be done later/separately, that's cool, just wondering if I missed something.

**Review by allisaurus — APPROVED** [review]
Had a question for my own understanding, but I am down with 1) running this as a status blocker on merges to master 2) running this as a non-blocking check on PRs. Thanks for drafting this!

Approving on the condition other folks with opinions on build/PR CI are also cool with things.

File: .github/workflows/integration_tests.yml
Comment (allisaurus) [inline .github/workflows/integration_tests.yml:7]: How are we specifying that this check is non-blocking to PRs? If that has to be done later/separately, that's cool, just wondering if I missed something.

Comment (german-muzquiz) [reply .github/workflows/integration_tests.yml:7]: PR status checks can be configured on the branch rules of the repo, under `Settings` page, there you can indicate which ones are required. For this repo you can see it in all PRs:

image

Also I see the following relevant rules of Mergify bot:

image

I guess those refer to the workflow job name, which is `build` for the PR build workflow and `it-test` for the integration tests workflow.

**Review by maggieneterval — APPROVED** [review]

**Review by jonsie — APPROVED** [review]

**Pull request merged**
